### PR TITLE
[build] fix warning of clearing non-trivial type

### PR DESCRIPTION
**Pull request by wgtdkp**
There is a warning with gcc/g++ 10.2.0 when clearing an object of non-trivial type with `memset`:

(build command: `cmake -GNinja -DOT_COMMISSIONER=ON .. && ninja`)

```shell
OPENTHREAD_CONFIG_THREAD_VERSION=OT_THREAD_VERSION_1_1
[65/132] Building CXX object src/core/CMakeFiles/openthread-ftd.dir/meshcop/commissioner.cpp.o
../src/core/meshcop/commissioner.cpp: In constructor ‘ot::MeshCoP::Commissioner::Commissioner(ot::Instance&)’:
../src/core/meshcop/commissioner.cpp:78:41: warning: ‘void* memset(void*, int, size_t)’ clearing an object of non-trivial type ‘struct ot::MeshCoP::Commissioner::Joiner’; use assignment or value-initialization instead [-Wclass-memaccess]
   78 |     memset(mJoiners, 0, sizeof(mJoiners));
      |                                         ^
In file included from ../src/core/meshcop/commissioner.cpp:34:
../src/core/meshcop/commissioner.hpp:347:12: note: ‘struct ot::MeshCoP::Commissioner::Joiner’ declared here
  347 |     struct Joiner
      |            ^~~~~~
[132/132] Linking CXX static library src/ncp/libopenthread-ncp-ftd.a
```

This PR fixes this issue:
1. remove the problematic `memset` usages;
2. turn on `OT_COMPILE_WARNING_AS_ERROR` for all CI builds;

### CODE REVIEW

**Review by librasungirl — APPROVED** [review]

**Comment by simonlingoogle** [comment]
Note that code size increased by about 300B. @wgtdkp

**Comment by wgtdkp** [comment]
> Note that code size increased by about 300B. @wgtdkp

Yes, we introduced a new default constructor.

File: src/core/common/time.hpp [hunk 1]
```diff
@@ -64,19 +64,13 @@ class Time
...
-    Time(void) {}
-
     /**
      * This constructor initializes a `Time` object with a given value.
      *
      * @param[in] aValue   The numeric time value to initialize the `Time` object.
      *
      */
-    explicit Time(uint32_t aValue) { SetValue(aValue); }
+    explicit Time(uint32_t aValue = 0) { SetValue(aValue); }
```
Comment (simonlingoogle) [inline src/core/common/time.hpp:73]: I think `Time` is indeed POD. But if the compiler can't understand it, maybe we should ignore this specific warning?

**Review by abtink — COMMENTED** [review]
I noticed this PR is increasing code size (ncp/cli ftd) by ~270 bytes (I guess from extra constructor calls).

Some suggestions below where I think we can fix/address the compiler warning and avoid the overhead.

File: src/core/common/time.hpp (hunk 1 above)
Comment (wgtdkp) [reply src/core/common/time.hpp]: okay, suppress the warning also works for me. updated.

File: src/core/meshcop/commissioner.cpp [hunk 2]
```diff
@@ -75,8 +75,6 @@ Commissioner::Commissioner(Instance &aInstance)
     , mJoinerCallback(nullptr)
     , mCallbackContext(nullptr)
 {
-    memset(mJoiners, 0, sizeof(mJoiners));
```
[the remark this answers is not in the document]
Comment (wgtdkp) [reply src/core/meshcop/commissioner.cpp:78]: @abtink Thanks for the suggestion. Revised.

File: src/core/thread/network_data_leader_ftd.cpp [hunk 3]
```diff
@@ -69,7 +69,11 @@ void Leader::Reset(void)
 {
     LeaderBase::Reset();
 
-    memset(mContextLastUsed, 0, sizeof(mContextLastUsed));
```
[the remark this answers is not in the document]
Comment (wgtdkp) [reply src/core/thread/network_data_leader_ftd.cpp:72]: Revised.

**Review by simonlingoogle — APPROVED** [review]
LGTM :+1:

**Review by abtink — APPROVED** [review]
LGTM. Thanks 👍

**Review by jwhui — APPROVED** [review]

**Pull request merged**